bitcoin/src/blockdata/script/owned.rs

```diff
@@ -87,6 +87,27 @@ impl ScriptBuf {
         let rw = Box::into_raw(self.0.into_boxed_slice()) as *mut Script;
         unsafe { Box::from_raw(rw) }
     }
+
+    /// Pushes `n` onto the script.
+    ///
+    /// Only meant for usage by the `bitcoin::script::ScriptBufExt` trait, consider using the API
+    /// provided by that trait.
+    #[doc(hidden)]
+    pub fn push(&mut self, n: u8) { self.0.push(n) }
+
+    /// Pops the last byte from the script if there is one.
+    ///
+    /// Only meant for usage by the `bitcoin::script::ScriptBufExt` trait, consider using the API
+    /// provided by that trait.
+    #[doc(hidden)]
+    pub fn pop(&mut self) -> Option<u8> { self.0.pop() }
+
+    /// Extends the script with bytes from `other`.
+    ///
+    /// Only meant for usage by the `bitcoin::script::ScriptBufExt` trait, consider using the API
+    /// provided by that trait.
+    #[doc(hidden)]
+    pub fn extend_from_slice(&mut self, other: &[u8]) { self.0.extend_from_slice(other) }
```

Comment:
I don't think we need to hide them but we likely can make the API somehow better.

The current design of `ScriptBuf` pretty much forces us to use `Vec` internally. There are direct conversions that promise not to reallocate, there are various other methods consistent with `Vec` and it's hard to imagine what other layout it could be. The only sensible alternatives that come to my mind are `Cow<'static, [u8]>` and its crazier version that uses `capacity = 0` to denote static data. However I doubt it's worth additional complexity. Scripts in Bitcoin are almost never constant - they almost always contain a unique key/signature and such is almost never baked into the memory. (The only notable exception being the Genesis block.)

Also the way we've structured our code makes it pretty easy to work with custom types. We already support `Box<Script>` and probably also `Cow<'_, Script>`. It's just whatever is in `Transaction` has fixed types and I think it's fine. With `AsTransaction` trait this won't be a big deal and we could add parameters to `Transaction` if we really want to later.

So I think I wouldn't be opposed to just having `as_byte_vec(&mut self) -> &mut Vec<u8>`. Yes, users will be able to do crazy stuff with it but that's already possible:

```rust
let mut tmp = core::mem::take(script).into_bytes(); // assuming the user only has &mut ScriptBuf
tmp.push(42);
*script = ScriptBuf::from_bytes(tmp);
```

Though this hack is also an argument why we don't need any of the methods at all. It may be hard for LLVM to optimize-out the swap but that can probably be hacked by having a guard put the vec back (so that all paths including panics keep the value the same).

So if you really really really want to be conservative about the layout and API I suggest you use the hack.